Key binding context

deirn opened the pull request:
Add `KeyBindingContext` to register multiple key binds that won't conflict with others.

This PR doesn't change vanilla behavior in which only one conflicting key (same context and key) will be registered as pressed. 
The order in which what key to functions depends on when the `KeyBinding` instance is created.

~~Do we want `KeyBinding#isPressed` and `wasPressed` to work when a screen is open?~~ Now it also makes it usable on screen.

deirn committed:
key binding context

deirn committed:
only press one conflicting key bind

deirn committed:
replace the conflict text

deirn committed:
fix checkstyle

deirn committed:
add license header

modmuss50 commented on fabric-key-binding-api-v1/src/client/java/net/fabricmc/fabric/api/client/keybinding/v1/KeyBindingHelper.java:
```diff
@@ -47,8 +47,20 @@ private KeyBindingHelper() {
 	 * @throws IllegalArgumentException when a key binding with the same ID is already registered
 	 */
 	public static KeyBinding registerKeyBinding(KeyBinding keyBinding) {
+		return registerKeyBinding(keyBinding, KeyBindingContext.IN_GAME);
```
Should this not be all for existing keybinds? Correct me if im wrong but can exitsing keybinds be used when in a screen?

deirn replied:
> Should this not be all for existing keybinds? 

Yes, that is the default context for all keybinds.

> Can exitsing keybinds be used when in a screen?

No, you need to check the bounded key if it's pressed manually. (eg. via `InputUtil.isKeyPressed` or on screen's key methods)

modmuss50 commented on fabric-key-binding-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/keybinding/KeyBindingEntryMixin.java:
```diff
@@ -0,0 +1,36 @@
+
+	@ModifyConstant(method = "update", constant = @Constant(stringValue = "controls.keybinds.duplicateKeybinds"))
+	private String replaceConflictText(String constant) {
+		return "fabric.keybinding.conflicts";
```
We shouldnt always be replacing this, it will make this text no longer translated for anyone using fabric API even if there are no mods using this new API.

deirn replied:
Done, it's now only changed if the conflicting key has a different context.

modmuss50 commented on fabric-key-binding-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/client/keybinding/CustomKeyBindingContext.java:
```diff
@@ -0,0 +1,28 @@
+
+import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingContext;
+
+public class CustomKeyBindingContext implements KeyBindingContext {
+	@Override
+	public boolean isActive() {
+		return MinecraftClient.getInstance().currentScreen == null;
+	}
+}
```
~~Is this not the same as IN_GAME? Might be better to have a test that does something slightly different.~~ Just this is is `isActive` and not `conflictsWith`. I think the test mod needs expanding to cover more cases.

Maybe also look into automated testing, but that likely requires me to finish: <https://github.com/FabricMC/fabric/pull/3168> Unit tests might be possible.

deirn replied:
This is to check whether 2 keys with different context _instance_ work simultaneously.

modmuss50 commented on fabric-key-binding-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/client/keybinding/KeyBindingsTest.java:
```diff
@@ -51,6 +66,30 @@ public void onInitializeClient() {
 				client.player.sendMessage(Text.literal("Duplicate Key was pressed!"), false);
 			}
+
+			while (inGameBinding.wasPressed()) {
```
A small comment saying how to test this might be helpful. Just list out a few situations and the expected outcome.

modmuss50 commented on fabric-key-binding-api-v1/src/client/java/net/fabricmc/fabric/api/client/keybinding/v1/KeyBindingContext.java:
```diff
@@ -0,0 +1,74 @@
+	 */
+	@ApiStatus.OverrideOnly
+	default boolean conflictsWith(KeyBindingContext other) {
+		return this == other;
```
This doesnt feel right for some reason. If I understand this correctly both keybinds must not conflict with each other. So evem if you did return false to have a keybind that could run everywhere it still wouldnt work.

modmuss50 commented on fabric-key-binding-api-v1/src/client/java/net/fabricmc/fabric/api/client/keybinding/v1/KeyBindingContext.java:
```diff
@@ -0,0 +1,74 @@
+		return left.conflictsWith(right) || right.conflictsWith(left);
+	}
+
+	boolean isActive();
```
Can you document this?

modmuss50 commented on fabric-key-binding-api-v1/src/client/java/net/fabricmc/fabric/api/client/keybinding/v1/KeyBindingHelper.java:
```diff
@@ -47,8 +47,20 @@ private KeyBindingHelper() {
+	 * @throws IllegalArgumentException when a key binding with the same ID is already registered
+	 */
+	public static KeyBinding registerKeyBinding(KeyBinding keyBinding, KeyBindingContext context) {
 		Objects.requireNonNull(keyBinding, "key binding cannot be null");
```
Also null check the context

modmuss50 reviewed (commented):
Im not fully convinced by the API design here, it feels a little brittle. I dont really have a good idea for something better atm.

deirn committed:
update javadoc
provide client instance on `isActive`
require explicit impl for `conflictsWith`

deirn committed:
skip solving conflicting key if it's only one

deirn committed:
add item context test

deirn committed:
conflictsWith -> conflicts

deirn committed:
make `KeyBinding#isPressed` and `wasPressed` work on screen

deirn committed:
only change conflict text if there's multiple conflicting context

deirn committed:
add null check

deirn committed:
fix style

deirn committed:
explain the test more

deirn committed:
actually only call one conflicting key

modmuss50 reviewed (approved):
API looks good to me, I think it makes more sense than when I last looked at it. The mixin looks a little friagile, but i dont think there is a lot that can be done about that.
